File: python/rotateArray.py
# Given an integer array nums, rotate the array to the right by k steps, where k is non-negative.

def rotate1(nums, k):
    copy = list(nums)
    for i in range(0, len(copy)):
        nums[(i+k)%len(nums)] = copy[i]

def rotate2(nums, k):
    k = k % len(nums)
    def reverse(start, end):
        while start < end:
            backup = nums[start]
            nums[start] = nums[end]
            nums[end] = backup
            start += 1
            end -= 1
    reverse(0, len(nums) - 1)
    reverse(0, k-1)
    reverse(k, len(nums) - 1)

File: leetCode.com/typescript/array/rotateArray.ts
// Given an integer array nums, rotate the array to the right by k steps, where k is non-negative.


function rotate1(nums: number[], k: number): void {
    let copy = Object.assign([], nums)
    for (let i = 0; i < copy.length; i++){
        nums[(i + k) % nums.length] = copy[i]
    }
};

function rotate2(nums: number[], k: number): void {
    k = k % nums.length
    function reverse(start: number, end: number): void{
        while (start < end) {
            [nums[start], nums[end]] = [nums[end], nums[start]]
            start++
            end--
        }
    }

    reverse(0, nums.length - 1)
    reverse(0, k - 1)
    reverse(k, nums.length -1)
}